language/vm/vm-runtime/vm-runtime-types/src/value.rs

```diff
@@ -131,12 +121,475 @@ pub struct RootAccessPath {
+#[derive(Debug, Clone)]
 pub struct GlobalRef {
+    val: MutVal,
     root: Rc<RefCell<RootAccessPath>>,
-    reference: MutVal,
+}
+
+impl Referenceable for GlobalRef {
+    fn borrow(&self) -> VMResult<VMRef> {
+        self.val.borrow()
+    }
+
+    fn borrow_mut(&self) -> VMResult<VMRefMut> {
+        self.root.borrow_mut().mark_dirty();
+        self.val.borrow_mut()
+    }
+}
+
+impl GlobalRef {
+    pub fn make_root(ap: AccessPath, value: Value) -> Self {
+        Self {
+            root: Rc::new(RefCell::new(RootAccessPath::new(ap))),
+            val: MutVal::new(value),
+        }
+    }
+
+    pub fn move_to(ap: AccessPath, value: Struct) -> Self {
+        let mut root = RootAccessPath::new(ap);
+        root.mark_dirty();
+        GlobalRef {
+            root: Rc::new(RefCell::new(root)),
+            val: MutVal::new(Value::struct_(value)),
+        }
+    }
+
+    pub fn move_from(&mut self) -> VMResult<Value> {
+        self.root.borrow_mut().mark_deleted();
+        self.val.copy_value()
+    }
+
+    pub fn is_loadable(&self) -> bool {
+        !self.is_deleted()
+    }
+
+    pub fn is_dirty(&self) -> bool {
+        self.root.borrow().status == GlobalDataStatus::DIRTY
+    }
+
+    pub fn is_deleted(&self) -> bool {
+        self.root.borrow().status == GlobalDataStatus::DELETED
+    }
+
+    pub fn is_clean(&self) -> bool {
+        self.root.borrow().status == GlobalDataStatus::CLEAN
+    }
+
+    // Return the resource behind the reference.
+    // If the reference is not exclusively held by the cache (ref count 0) returns None
+    pub fn get_data(self) -> Option<Value> {
+        match Rc::try_unwrap(self.root) {
+            Ok(_) => match Rc::try_unwrap(self.val.0) {
+                Ok(res) => Some(Value::new(res.into_inner())),
+                Err(_) => None,
+            },
+            Err(_) => None,
+        }
+    }
+
+    pub fn size(&self) -> AbstractMemorySize<GasCarrier> {
+        words_in(*REFERENCE_SIZE)
+    }
+
+    fn pretty_string(&self) -> String {
+        format!(
+            "GlobalRef(({}, {}), {})",
+            Rc::strong_count(&self.root),
+            self.root.borrow().pretty_string(),
+            self.val.pretty_string()
+        )
+    }
+}
+
+/// A reference that holds a shared pointer to the value being referenced directly.
+///
+/// `DirectRef` is the only kind of reference that can be extended -- a.k.a creating
+/// a reference to a sub component of the value referenced. For example, borrowing a
+/// field of a struct or borrowing an element of a vector.
+///
+/// This means all other references in the VM are indirect -- they all contain a
+/// `DirectRef` and some auxiliary info (e.g. an index) to further locate the data
+/// behind the shared pointer.
+#[derive(Debug, Clone)]
+pub enum DirectRef {
+    Local(LocalRef),
+    Global(GlobalRef),
+}
+
+impl DirectRef {
+    // TODO: revisit.
+    pub fn from_value_for_cost_synthesis(v: Value) -> Self {
```

Comment:
@tzakian Does this look good to you? On one hand we have cost_synthesis creating various values, and on the other hand the old philosophy seemed to be that values should be created inside this module and behave as black boxes to the outside world.